### ISSUE TITLE
Agent in Not Connected Status after reboot CMR

### ISSUE THREAD

**Issue by yvbondarenko** [inspectIT/inspectIT#227]
Hi, I have same problem with agent when I try to restart CMR.
I have instrumented Apache Tomcat 8.
Afrer reboot CMR I navigate to few link in Apache Tomcat 8.

```
*******************************************************************
Agent Log:
2017-02-01 08:18:25,239: 1161   [           main] INFO  nfig.impl.ConfigurationStorage - Agent name set to: SB-Apache8
2017-02-01 08:18:25,328: 1250   [           main] INFO  izer.schema.ClassSchemaManager - ||-Class Schema Manager started..
2017-02-01 08:18:25,719: 1641   [           main] INFO  nection.impl.KryoNetConnection - KryoNet: Connecting to 10.49.197.61:9070
2017-02-01 08:18:25,788: 1710   [           main] INFO  nection.impl.KryoNetConnection - KryoNet: Connection established!
2017-02-01 08:18:26,044: 1966   [           main] INFO  nfig.impl.ConfigurationStorage - Agent configuration added with following configuration interface properties:
2017-02-01 08:18:26,045: 1967   [           main] INFO  nfig.impl.ConfigurationStorage - Assigned environment: Default Environment
2017-02-01 08:18:26,045: 1967   [           main] INFO  nfig.impl.ConfigurationStorage - Active profiles:
2017-02-01 08:18:26,045: 1967   [           main] INFO  nfig.impl.ConfigurationStorage - |-[Common] Exclude Classes
2017-02-01 08:18:26,045: 1967   [           main] INFO  nfig.impl.ConfigurationStorage - |-[Common] HTTP
2017-02-01 08:18:26,045: 1967   [           main] INFO  nfig.impl.ConfigurationStorage - |-[Common] SQL
2017-02-01 08:18:26,045: 1967   [           main] INFO  nfig.impl.ConfigurationStorage - Options:
2017-02-01 08:18:26,045: 1967   [           main] INFO  nfig.impl.ConfigurationStorage - |-class loading delegation: true
2017-02-01 08:18:26,045: 1967   [           main] INFO  nfig.impl.ConfigurationStorage - |-enhanced exception sensor: true
2017-02-01 08:18:26,046: 1968   [           main] INFO  nfig.impl.ConfigurationStorage - Class-cache exists on the server: true
2017-02-01 08:18:26,046: 1968   [           main] INFO  nfig.impl.ConfigurationStorage - Number of initially instrumented classes: 12
2017-02-01 08:18:26,256: 2178   [           main] INFO  spectit.agent.java.SpringAgent - Spring successfully initialized
2017-02-01 08:18:26,257: 2179   [           main] INFO  spectit.agent.java.SpringAgent - Using agent version 1.7.7.90.
2017-02-01 08:19:11,244: 47166  [ervice-thread-2] INFO  nnection.impl.KeepAliveManager - Trying to reconnect to the server.
2017-02-01 08:19:16,244: 52166  [ervice-thread-2] INFO  nnection.impl.KeepAliveManager - Trying to reconnect to the server.
2017-02-01 08:19:26,246: 62168  [ervice-thread-1] INFO  nnection.impl.KeepAliveManager - Trying to reconnect to the server.
2017-02-01 08:19:46,245: 82167  [ervice-thread-2] INFO  nnection.impl.KeepAliveManager - Trying to reconnect to the server.
2017-02-01 08:20:26,246: 122168 [ervice-thread-1] INFO  nnection.impl.KeepAliveManager - Trying to reconnect to the server.
2017-02-01 08:21:46,248: 202170 [ervice-thread-0] INFO  nnection.impl.KeepAliveManager - Trying to reconnect to the server.
2017-02-01 08:21:46,312: 202234 [ervice-thread-0] INFO  nection.impl.KryoNetConnection - KryoNet: Reconnection successful!
2017-02-01 08:21:59,129: 215051 [nio-6060-exec-8] ERROR analyzer.impl.ByteCodeAnalyzer - Error occurred instrumenting the byte code of class java.text.ParsePosition
rocks.inspectit.shared.all.exception.BusinessException: The 'Agent Does Not Exist' exception occurred on the 'Agent Management' component. Exception occurred executing the action: 'Instrumenting class with hash '28c0ccab9452e6ce2cf3b1df78b2411ec21838261bde5119a1068b53b51e1b92' for the agent with id=4752'.
Wrapped by: java.util.concurrent.ExecutionException: rocks.inspectit.shared.all.exception.BusinessException: The 'Agent Does Not Exist' exception occurred on the 'Agent Management' component. Exception occurred executing the action: 'Instrumenting class with hash '28c0ccab9452e6ce2cf3b1df78b2411ec21838261bde5119a1068b53b51e1b92' for the agent with id=4752'.
	at java.util.concurrent.FutureTask.report(Unknown Source)
	at java.util.concurrent.FutureTask.get(Unknown Source)
	at rocks.inspectit.agent.java.analyzer.impl.ByteCodeAnalyzer.analyzeAndInstrumentInternal(ByteCodeAnalyzer.java:198)
	at rocks.inspectit.agent.java.analyzer.impl.ByteCodeAnalyzer.analyzeAndInstrument(ByteCodeAnalyzer.java:134)
	at rocks.inspectit.agent.java.SpringAgent.inspectByteCode(SpringAgent.java:215)
	at rocks.inspectit.agent.java.javaagent.JavaAgent.transform(JavaAgent.java:153)
	at sun.instrument.TransformerManager.transform(Unknown Source)
	at sun.instrument.InstrumentationImpl.transform(Unknown Source)
	at java.text.DateFormat.parse(Unknown Source)
	at org.apache.tomcat.util.http.FastHttpDateFormat.internalParseDate(FastHttpDateFormat.java:178)
	at org.apache.tomcat.util.http.FastHttpDateFormat.parseDate(FastHttpDateFormat.java:157)
	at org.apache.catalina.connector.Request.getDateHeader(Request.java:2131)
	at org.apache.catalina.connector.RequestFacade.getDateHeader(RequestFacade.java:678)
	at org.apache.catalina.servlets.DefaultServlet.checkIfModifiedSince(DefaultServlet.java:1994)
	at org.apache.catalina.servlets.DefaultServlet.checkIfHeaders(DefaultServlet.java:687)
	at org.apache.catalina.servlets.DefaultServlet.serveResource(DefaultServlet.java:806)
	at org.apache.catalina.servlets.DefaultServlet.doGet(DefaultServlet.java:425)
	at javax.servlet.http.HttpServlet.service(Unknown Source)
	at javax.servlet.http.HttpServlet.service(Unknown Source)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:230)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:165)
	at org.apache.tomcat.websocket.server.WsFilter.doFilter(Unknown Source)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:192)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:165)
	at org.apache.catalina.filters.SetCharacterEncodingFilter.doFilter(Unknown Source)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:192)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:165)
	at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:198)
	at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:96)
	at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:474)
	at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:140)
	at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:79)
	at org.apache.catalina.valves.AbstractAccessLogValve.invoke(AbstractAccessLogValve.java:624)
	at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:87)
	at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:349)
	at org.apache.coyote.http11.Http11Processor.service(Http11Processor.java:783)
	at org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:66)
	at org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:798)
	at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1434)
	at org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:49)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
	at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)
	at java.lang.Thread.run(Unknown Source)
2017-02-01 08:21:59,136: 215058 [nio-6060-exec-8] ERROR analyzer.impl.ByteCodeAnalyzer - Error occurred instrumenting the byte code of class java.text.CalendarBuilder
rocks.inspectit.shared.all.exception.BusinessException: The 'Agent Does Not Exist' exception occurred on the 'Agent Management' component. Exception occurred executing the action: 'Instrumenting class with hash '7a5b783f1addfed123675e9a72ee7ffbf90d9dad91e521066a089c8368c97c8c' for the agent with id=4752'.
Wrapped by: java.util.concurrent.ExecutionException: rocks.inspectit.shared.all.exception.BusinessException: The 'Agent Does Not Exist' exception occurred on the 'Agent Management' component. Exception occurred executing the action: 'Instrumenting class with hash '7a5b783f1addfed123675e9a72ee7ffbf90d9dad91e521066a089c8368c97c8c' for the agent with id=4752'.
	at java.util.concurrent.FutureTask.report(Unknown Source)
	at java.util.concurrent.FutureTask.get(Unknown Source)
	at rocks.inspectit.agent.java.analyzer.impl.ByteCodeAnalyzer.analyzeAndInstrumentInternal(ByteCodeAnalyzer.java:198)
	at rocks.inspectit.agent.java.analyzer.impl.ByteCodeAnalyzer.analyzeAndInstrument(ByteCodeAnalyzer.java:134)
	at rocks.inspectit.agent.java.SpringAgent.inspectByteCode(SpringAgent.java:215)
	at rocks.inspectit.agent.java.javaagent.JavaAgent.transform(JavaAgent.java:153)
	at sun.instrument.TransformerManager.transform(Unknown Source)
	at sun.instrument.InstrumentationImpl.transform(Unknown Source)
	at java.text.SimpleDateFormat.parse(Unknown Source)
	at java.text.DateFormat.parse(Unknown Source)
	at org.apache.tomcat.util.http.FastHttpDateFormat.internalParseDate(FastHttpDateFormat.java:178)
	at org.apache.tomcat.util.http.FastHttpDateFormat.parseDate(FastHttpDateFormat.java:157)
	at org.apache.catalina.connector.Request.getDateHeader(Request.java:2131)
	at org.apache.catalina.connector.RequestFacade.getDateHeader(RequestFacade.java:678)
	at org.apache.catalina.servlets.DefaultServlet.checkIfModifiedSince(DefaultServlet.java:1994)
	at org.apache.catalina.servlets.DefaultServlet.checkIfHeaders(DefaultServlet.java:687)
	at org.apache.catalina.servlets.DefaultServlet.serveResource(DefaultServlet.java:806)
	at org.apache.catalina.servlets.DefaultServlet.doGet(DefaultServlet.java:425)
	at javax.servlet.http.HttpServlet.service(Unknown Source)
	at javax.servlet.http.HttpServlet.service(Unknown Source)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:230)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:165)
	at org.apache.tomcat.websocket.server.WsFilter.doFilter(Unknown Source)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:192)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:165)
	at org.apache.catalina.filters.SetCharacterEncodingFilter.doFilter(Unknown Source)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:192)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:165)
	at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:198)
	at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:96)
	at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:474)
	at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:140)
	at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:79)
	at org.apache.catalina.valves.AbstractAccessLogValve.invoke(AbstractAccessLogValve.java:624)
	at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:87)
	at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:349)
	at org.apache.coyote.http11.Http11Processor.service(Http11Processor.java:783)
	at org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:66)
	at org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:798)
	at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1434)
	at org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:49)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
	at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)
	at java.lang.Thread.run(Unknown Source)
..............................
```

**Comment by ivansenic** [comment]
Hi @yvbondarenko,

Yes this is the bug we are aware of. After restarting the CMR the agent reconnects to the server, but it functionality is limited wrt dynamic instrumentation and it's marked as disconnected. However, sending of the data does work after re-connection. I do understand that this is confusing completely. I will open a ticket for this bug and we'll add it to our next sprint.

**Comment by ivansenic** [comment]
Ticket for this bug: [INSPECTIT-2322](https://inspectit-performance.atlassian.net/browse/INSPECTIT-2322)

**Comment by yvbondarenko** [comment]
Thanks a lot

**Comment by youjenli** [comment]
Hi @ivansenic 

What is the limitation (eg. storage capacity) of this feature?  
I think the java agent may not know how long will the CMR be offline.  
Will it keep saving metrics locally until it exceed hardware limitation (eg. out of memory, out of disk space)?

**Comment by ivansenic** [comment]
Hi @youjenli,

Metrics are by default sent every 5 seconds to the CMR. If the CMR is offline, there will be (by default) 3 re-tries. After that metrics are dropped and they won't be sent. So out of memory and similar limitations should not be reached, but the data is effectively lost.

**Comment by ivansenic** [comment]
Fixed by #326
